File: Backend/appointment_module/src/main/java/com/example/demo/services/AppointmentService.java
package com.example.demo.services;


import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.dao.EmptyResultDataAccessException;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.stereotype.Service;

import com.example.demo.models.Appointment;
import com.example.demo.repositories.AppointmentRepository;

import java.util.List;
import java.util.Optional;

@Service
public class AppointmentService {

	 

    @Autowired
    private AppointmentRepository appointmentRepository;

    // Book an appointment
    public Appointment bookAppointment(Appointment appointment) {
        appointment.setStatus("Pending");
        return appointmentRepository.save(appointment);
    }
//    public Appointment bookAppointment(Appointment appointment, String email) {
//        // ✅ Fetch `userId` from the database using the logged-in user's email
//        String sql = "SELECT id FROM users WHERE email = ?";
//        Long userId;
//
//        try {
//            userId = jdbcTemplate.queryForObject(sql, Long.class, email);
//        } catch (EmptyResultDataAccessException e) {
//            throw new RuntimeException("User not found for email: " + email);
//        }
//
//        appointment.setUserId(userId); // ✅ Assign `userId`
//        appointment.setStatus("Pending");
//
//        return appointmentRepository.save(appointment);
//    }



    // Get appointments by user
    public List<Appointment> getAppointmentsByUser(Long userId) {
        return appointmentRepository.findByUserId(userId);
    }

    // Get appointments by doctor
    public List<Appointment> getAppointmentsByDoctor(Long doctorId) {
        return appointmentRepository.findByDoctorId(doctorId);
    }

    // Update appointment status
    public Appointment updateAppointmentStatus(Long id, String status) {
        Optional<Appointment> optionalAppointment = appointmentRepository.findById(id);
        if (optionalAppointment.isPresent()) {
            Appointment appointment = optionalAppointment.get();
            appointment.setStatus(status);
            return appointmentRepository.save(appointment);
        }
        return null;
    }
    public boolean updateAppointmentStatus(Long id, Long doctorId, String prescription) {
        Appointment appointment = appointmentRepository.findById(id).orElse(null);

        if (appointment == null) {
            System.out.println("Error: Appointment ID " + id + " not found.");
            return false;
        }

        // ✅ Ensure the doctor is assigned to this appointment
        if (!appointment.getDoctorId().equals(doctorId)) {
            System.out.println("Error: Doctor ID " + doctorId + " does not match appointment's assigned doctor.");
            return false;
        }

        appointment.setStatus(prescription); // 🔥 Store prescription in status
        appointmentRepository.save(appointment);

        System.out.println("Appointment updated successfully: " + appointment);
        return true;
    }


    // Cancel appointment
    public boolean cancelAppointment(Long id) {
        if (appointmentRepository.existsById(id)) {
            appointmentRepository.deleteById(id);
            return true;
        }
        return false;
    }
}

File: doctor-patient-portal/src/services/AppointmentService.js
import axios from "axios";

const API_URL = "http://localhost:8888/appointments"; // Change this to your backend URL

// Book an appointment
export const bookAppointment = async (appointmentData) => {
  return axios.post(`${API_URL}/user/book`, appointmentData);
};

// Get appointments by user
export const getAppointmentsByUser = async (userId) => {
    if (!userId) {
      console.error("UserId is missing in API call!");
      return { data: [] };
    }
    console.log(`Making API call: ${API_URL}/user/${userId}`);
    return axios.get(`${API_URL}/user/${userId}`);
  };
  

export const getAppointmentsByDoctor = async (doctorId) => {
    try {
      console.log(`Calling API: ${API_URL}/doctor/${doctorId}`);
      const response = await axios.get(`${API_URL}/doctor/${doctorId}`);
      return response;
    } catch (error) {
      console.error("Error fetching appointments for doctor:", error);
      throw error;
    }
}

// Update appointment status
export const updateAppointmentStatus = async (id, status) => {
  return axios.put(`${API_URL}/${id}/update-status`, { status });
};

// Cancel appointment
export const cancelAppointment = async (id) => {
  return axios.delete(`${API_URL}/cancel/${id}`);
};
